tortoisebeacon/weakcoin/weak_coin_test.go

```diff
@@ -1,118 +1,378 @@
+				Round:     round,
+				Unit:      1,
+				Signature: zeroLSB,
+			}},
+		},
+		{
+			desc:         "PreviousEpoch",
+			local:        zeroLSB,
+			allowances:   weakcoin.UnitAllowances{string(zeroLSB): 1, string(oneLSB): 1},
+			startedEpoch: epoch,
+			startedRound: round,
+			messages: []weakcoin.Message{{
+				Epoch:     epoch - 1,
+				Round:     round,
+				Unit:      1,
+				Signature: oneLSB,
+			}},
 		},
 		{
-			name:      "Case 2",
-			proposal:  proposal2.Bytes(),
-			threshold: threshold.Bytes(),
-			expected:  false,
+			desc:         "PreviousRound",
+			local:        zeroLSB,
+			allowances:   weakcoin.UnitAllowances{string(zeroLSB): 1, string(oneLSB): 1},
+			startedEpoch: epoch,
+			startedRound: round,
+			messages: []weakcoin.Message{{
+				Epoch:     epoch,
+				Round:     round - 1,
+				Unit:      1,
+				Signature: oneLSB,
+			}},
 		},
 	}
+	for _, tc := range tcs {
+		tc := tc
+		t.Run(tc.desc, func(t *testing.T) {
+			wc := weakcoin.New(
+				noopBroadcaster(t, ctrl),
+				staticSigner(t, ctrl, tc.local), verifier,
+				weakcoin.WithConfig(config))
+
+			wc.StartEpoch(tc.startedEpoch, tc.allowances)
+			require.NoError(t, wc.StartRound(context.TODO(), tc.startedRound))
+
+			for _, msg := range tc.messages {
+				wc.HandleSerializedMessage(context.TODO(), broadcastedMessage(t, ctrl, msg), nil)
+			}
+			wc.CompleteRound()
+
+			require.Equal(t, tc.coinflip, wc.Get(tc.startedEpoch, tc.startedRound))
+		})
+	}
+
+	for _, tc := range tcs {
+		tc := tc
+		t.Run("BufferingStartEpoch/"+tc.desc, func(t *testing.T) {
+			wc := weakcoin.New(
+				noopBroadcaster(t, ctrl),
+				staticSigner(t, ctrl, tc.local), verifier,
+				weakcoin.WithConfig(config))
+
+			wc.StartEpoch(tc.startedEpoch, tc.allowances)
+
+			for _, msg := range tc.messages {
+				wc.HandleSerializedMessage(context.TODO(), broadcastedMessage(t, ctrl, msg), nil)
+			}
+			require.NoError(t, wc.StartRound(context.TODO(), tc.startedRound))
+			wc.CompleteRound()
+
+			require.Equal(t, tc.coinflip, wc.Get(tc.startedEpoch, tc.startedRound))
+		})
+	}
+
+	for _, tc := range tcs {
+		tc := tc
+		t.Run("BufferingNextRound/"+tc.desc, func(t *testing.T) {
+			wc := weakcoin.New(
+				noopBroadcaster(t, ctrl),
+				staticSigner(t, ctrl, tc.local), verifier,
+				weakcoin.WithConfig(config))
+
+			wc.StartEpoch(tc.startedEpoch, tc.allowances)
+			require.NoError(t, wc.StartRound(context.TODO(), tc.startedRound))
+
+			for _, msg := range tc.messages {
+				msg.Round++
+				wc.HandleSerializedMessage(context.TODO(), broadcastedMessage(t, ctrl, msg), nil)
+			}
 
-	for _, tc := range tt {
+			require.NoError(t, wc.StartRound(context.TODO(), tc.startedRound+1))
+			wc.CompleteRound()
+
+			require.Equal(t, tc.coinflip, wc.Get(tc.startedEpoch, tc.startedRound+1))
+			wc.CompleteEpoch()
+		})
+	}
+	for _, tc := range tcs {
 		tc := tc
-		t.Run(tc.name, func(t *testing.T) {
-			t.Parallel()
+		t.Run("BufferingNextEpochAfterCompletion/"+tc.desc, func(t *testing.T) {
```

Comment:
`t.Parallel()`

Reply:
i created single mocking controller instance. i don't know if it will result in any concurrency issues. but the difference in run time is nelgibile. 0.110 s vs 0.100 s

Reply:
Let's leave it as it is then